clients/fides-js/src/lib/gpp/string-to-consent.ts

```diff
@@ -0,0 +1,308 @@
+import { CmpApi, SignalStatus, TcfEuV2, TcfEuV2Field } from "@iabgpp/cmpapi";
+
+import {
+  ComponentType,
+  ConsentMethod,
+  FidesCookie,
+  NoticeConsent,
+  PrivacyExperience,
+  PrivacyExperienceMinimal,
+  SaveConsentPreference,
+} from "../consent-types";
+import {
+  constructFidesRegionString,
+  createConsentPreferencesToSave,
+} from "../consent-utils";
+import { decodeFidesString } from "../fidesString";
+import { areLocalesEqual } from "../i18n/i18n-utils";
+import { updateConsentPreferences } from "../preferences";
+import { EMPTY_ENABLED_IDS } from "../tcf/constants";
+import { EnabledIds, TcfSavePreferences } from "../tcf/types";
+import {
+  createTCFConsentPreferencesToSave,
+  createTcfSavePayload,
+  createTcfSavePayloadFromMinExp,
+  updateTCFCookie,
+} from "../tcf/utils";
+import { FIDES_US_REGION_TO_GPP_SECTION } from "./constants";
+import { GPPUSApproach } from "./types";
+import { deriveGppFieldRegion, US_NATIONAL_REGION } from "./us-notices";
+
+const FORBIDDEN_LEGITIMATE_INTEREST_PURPOSE_IDS = [1, 3, 4, 5, 6];
+
+/**
+ * Gets the consent object from a GPP CMP API model based on the experience configuration
+ *
+ * This is the inverse of setGppOptOutsFromCookieAndExperience
+ */
+const getConsentFromGppCmpApi = ({
+  cmpApi,
+  experience,
+}: {
+  cmpApi: CmpApi;
+  experience: PrivacyExperience;
+}): NoticeConsent => {
+  const consent: NoticeConsent = {};
+  const {
+    privacy_notices: notices = [],
+    region: experienceRegion,
+    gpp_settings: gppSettings,
+  } = experience;
+  const usApproach = gppSettings?.us_approach;
+  let gppRegion = deriveGppFieldRegion({
+    experienceRegion,
+    usApproach,
+  });
+  let gppSection = FIDES_US_REGION_TO_GPP_SECTION[gppRegion];
+
+  if (!gppSection && usApproach === GPPUSApproach.ALL) {
+    // if we're using the all approach, and the current state isn't supported, we should default to national
+    gppRegion = US_NATIONAL_REGION;
+    gppSection = FIDES_US_REGION_TO_GPP_SECTION[gppRegion];
+  }
+
+  if (
+    !gppSection ||
+    (gppRegion === US_NATIONAL_REGION && usApproach === GPPUSApproach.STATE)
+  ) {
+    // If we don't have a section, we can't get anything.
+    // If we're using the state approach, we shouldn't use the national section, even if region is set to national.
+    return consent;
+  }
+
+  notices.forEach((notice) => {
```

Comment:
At this point, `gppSection` could still be undefined right? What do we want to do in this case?

Reply:
We `return consent;` in the line above so we don't reach this spot in that case.